### DISCUSSION TITLE
How to deploy my Express.js app?

### DISCUSSION THREAD

**Discussion by angel-luis** [vercel/vercel#8692] [Help] [answered]
I used to use Heroku, this is my `package.json` config:

```
  "scripts": {
    "start": "npm run build && node ./dist/services/startServer",
    "dev": "nodemon --exec npm run restart",
    "restart": "rimraf dist && npm run start",
    "build": "babel src -d dist --copy-files",
    "heroku-postbuild": "npm run build"
  },
```

The `import { createServer } from "http";` is located in `src/services/startServer.js` .

It throws an error: `Error: No Output Directory named "public" found after the Build completed. You can configure the Output Directory in your Project Settings.`

I don't have public folder because I don't need it, and I don't know how to properly configure the `vercel.json` file.

**Comment by swarnava** [comment, marked as answer]
Hi,

You can deploy APIs using pages/api in Next.js → https://nextjs.org/docs/api-routes/introduction.

See also: https://vercel.com/guides/using-express-with-vercel

The general rule is not to deploy an entire, monolithic Node.js/Express.js server on a pages/api which is a Serverless Function/AWS Lambda underneath. Serverless Functions are not meant to take on the load of an entire Node.js server.

**Reply by angel-luis** [reply]
Hi, really I don't want to use NextJS in that app, I just want an Express.js server as replacement of Heroku.

Following the guide, I've tried it creating a `vercel.json` config with this config, but it throws the same public folder error:
```
{
  "rewrites": [{ "source": "/api/(.*)", "destination": "/api" }]
}
```

**Reply by amyegan** [reply]
No problem if you prefer not to use Next.js in your app! Be sure you have added and committed the [`/public` directory](https://vercel.com/guides/using-express-with-vercel#adding-a-public-directory). If you have already done this and are still seeing the error on new deployments, please share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) so we can investigate what is going wrong. 🙏 

As Swarnava mentioned, it is generally recommended that you break up monolithic Express server into several smaller [Serverless Functions](https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions) as there are some [drawbacks and edge cases](https://vercel.com/guides/using-express-with-vercel#drawbacks-and-edge-cases) to consider if you deploy an Express.js application as a single Serverless Function. 

Serverless Functions do not require Next.js and can be built with one of a few [supported languages](https://vercel.com/docs/concepts/functions/serverless-functions/supported-languages) (including Node.js):

![image](https://user-images.githubusercontent.com/8247105/194595333-902a9f20-85c4-4583-afef-5b52f5c69dc2.png)

**Reply by angel-luis** [reply]
Hi, good monday!

I've created an empty `public` folder (with an empty file for Git tracking) and it worked!

The problem now is that all routes returns a 404 error.

The folder structure is:
![image](https://user-images.githubusercontent.com/17672548/194823555-054d4fa7-8bba-45a5-886a-e47d31cf2a69.png)

startServer.js
```javascript
#!/usr/bin/env node

/**
 * Module dependencies.
 */

import app from "../app";
import debug from "debug";
import { createServer } from "http";

/**
 * Get port from environment and store in Express.
 */

const port = normalizePort(process.env.PORT || "7777");
app.set("port", port);

/**
 * Create HTTP server.
 */

const server = createServer(app);

/**
 * Listen on provided port, on all network interfaces.
 */

server.listen(port);
server.on("error", onError);
server.on("listening", onListening);

/**
 * Normalize a port into a number, string, or false.
 */

function normalizePort(val) {
  const port = parseInt(val, 10);

  if (isNaN(port)) {
    // named pipe
    return val;
  }

  if (port >= 0) {
    // port number
    return port;
  }

  return false;
}

/**
 * Event listener for HTTP server "error" event.
 */

function onError(error) {
  if (error.syscall !== "listen") {
    throw error;
  }

  const bind = typeof port === "string" ? "Pipe " + port : "Port " + port;

  // handle specific listen errors with friendly messages
  switch (error.code) {
    case "EACCES":
      console.error(bind + " requires elevated privileges");
      process.exit(1);
      break;
    case "EADDRINUSE":
      console.error(bind + " is already in use");
      process.exit(1);
      break;
    default:
      throw error;
  }
}

/**
 * Event listener for HTTP server "listening" event.
 */

function onListening() {
  const addr = server.address();
  const bind = typeof addr === "string" ? "pipe " + addr : "port " + addr.port;
  debug("Listening on " + bind);
}
```

app.js
```javascript
// Transpile async/await
import "core-js/stable";
import "regenerator-runtime/runtime";
// Express itself
import express from "express";
// Express modules
import createError from "http-errors";
import path from "path";
import cookieParser from "cookie-parser";
import logger from "morgan";
import cors from "cors";
// Routes
import routes from "./api/index";
// Dotenv
import dotenv from "dotenv";
dotenv.config();

const app = express();

app.use(cors());
app.use(logger("dev"));
app.use(express.json({ limit: 3000000 }));
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, "public")));

app.use("/api", routes());

// catch 404 and forward to error handler
app.use(function (req, res, next) {
  next(createError(404));
});

// error handler
app.use(function (err, req, res, next) {
  console.log(err.message);
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get("env") === "development" ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.json(err.message);
});

export default app;
```

I can't migrate to serverless functions at this moment, I don't know if it's possible to deploy the app as an standard Express.js app or if I've to use another service.

Thank you for the support.

**Reply by angel-luis** [reply]
A few minutes later after deployment I got this error: `No more than 12 Serverless Functions can be added to a Deployment on the Hobby plan. Create a team (Pro plan) to deploy more. Learn More: https://vercel.link/function-count-limit` so I already know the problem!

In this case I will accept the later response as answer, thank you 🙂
